src/crimson/osd/osd.cc

```diff
@@ -580,57 +595,65 @@ seastar::future<Ref<PG>> OSD::load_pg(spg_t pgid)
-  case MSG_OSD_PG_PULL:
-    [[fallthrough]];
-  case MSG_OSD_PG_PUSH:
-    [[fallthrough]];
-  case MSG_OSD_PG_PUSH_REPLY:
-    [[fallthrough]];
-  case MSG_OSD_PG_RECOVERY_DELETE:
-    [[fallthrough]];
-  case MSG_OSD_PG_RECOVERY_DELETE_REPLY:
-    return handle_recovery_subreq(conn, boost::static_pointer_cast<MOSDFastDispatchOp>(m));
-  case MSG_OSD_PG_LEASE:
-    [[fallthrough]];
-  case MSG_OSD_PG_LEASE_ACK:
-    [[fallthrough]];
-  case MSG_OSD_PG_NOTIFY2:
-    [[fallthrough]];
-  case MSG_OSD_PG_INFO2:
-    [[fallthrough]];
-  case MSG_OSD_PG_QUERY2:
-    [[fallthrough]];
-  case MSG_OSD_BACKFILL_RESERVE:
-    [[fallthrough]];
-  case MSG_OSD_RECOVERY_RESERVE:
-    [[fallthrough]];
-  case MSG_OSD_PG_LOG:
-    return handle_peering_op(conn, boost::static_pointer_cast<MOSDPeeringOp>(m));
-  case MSG_OSD_REPOP:
-    return handle_rep_op(conn, boost::static_pointer_cast<MOSDRepOp>(m));
-  case MSG_OSD_REPOPREPLY:
-    return handle_rep_op_reply(conn, boost::static_pointer_cast<MOSDRepOpReply>(m));
-  default:
-    logger().info("{} unhandled message {}", __func__, *m);
-    return seastar::now();
-  }
+  return [this, conn, m] {
+    if (state.is_stopping()) {
+      return seastar::now();
+    }
+    switch (m->get_type()) {
+    case CEPH_MSG_OSD_MAP:
+      return handle_osd_map(conn, boost::static_pointer_cast<MOSDMap>(m));
+    case CEPH_MSG_OSD_OP:
+      return handle_osd_op(conn, boost::static_pointer_cast<MOSDOp>(m));
+    case MSG_OSD_PG_CREATE2:
+      shard_services.start_operation<CompoundPeeringRequest>(
+	*this,
+	conn->get_shared(),
+	m);
+      return seastar::now();
+    case MSG_COMMAND:
+      return handle_command(conn, boost::static_pointer_cast<MCommand>(m));
+    case MSG_OSD_PG_PULL:
+      [[fallthrough]];
+    case MSG_OSD_PG_PUSH:
+      [[fallthrough]];
+    case MSG_OSD_PG_PUSH_REPLY:
+      [[fallthrough]];
+    case MSG_OSD_PG_RECOVERY_DELETE:
+      [[fallthrough]];
+    case MSG_OSD_PG_RECOVERY_DELETE_REPLY:
+      return handle_recovery_subreq(conn, boost::static_pointer_cast<MOSDFastDispatchOp>(m));
+    case MSG_OSD_PG_LEASE:
+      [[fallthrough]];
+    case MSG_OSD_PG_LEASE_ACK:
+      [[fallthrough]];
+    case MSG_OSD_PG_NOTIFY2:
+      [[fallthrough]];
+    case MSG_OSD_PG_INFO2:
+      [[fallthrough]];
+    case MSG_OSD_PG_QUERY2:
+      [[fallthrough]];
+    case MSG_OSD_BACKFILL_RESERVE:
+      [[fallthrough]];
+    case MSG_OSD_RECOVERY_RESERVE:
+      [[fallthrough]];
+    case MSG_OSD_PG_LOG:
+      return handle_peering_op(conn, boost::static_pointer_cast<MOSDPeeringOp>(m));
+    case MSG_OSD_REPOP:
+      return handle_rep_op(conn, boost::static_pointer_cast<MOSDRepOp>(m));
+    case MSG_OSD_REPOPREPLY:
+      return handle_rep_op_reply(conn, boost::static_pointer_cast<MOSDRepOpReply>(m));
+    default:
+      logger().info("{} unhandled message {}", __func__, *m);
+      return seastar::now();
+    }
+  }().handle_exception([conn, m](std::exception_ptr eptr) {
```

Comment:
can we catch and ignore this exception in messenger ?

also, why not
```c++
handle_exception_type([] (crimson::common::system_shutdown_exception&) {
```
?

Reply:
> can we catch and ignore this exception in messenger ?

It is possible, but I think the major purpose to return a future from `ms_dispatch()` is to make throttling possible with `SocketConnection`.